engine/classes/Elgg/Upgrades/PrivateAndFriendAcls.php

```diff
@@ -0,0 +1,199 @@
+			$this->db->updateData($pmsql);
+
+			$fmsql = "UPDATE {$dbprefix}metadata SET access_id = {$friends_acl_id} WHERE access_id = -2 AND owner_guid = {$user->guid}";
+			$this->db->updateData($fmsql);
+
+			$pasql = "UPDATE {$dbprefix}annotations SET access_id = {$private_acl_id} WHERE access_id = 0 AND owner_guid = {$user->guid}";
+			$this->db->updateData($pasql);
+
+			$fasql = "UPDATE {$dbprefix}annotations SET access_id = {$friends_acl_id} WHERE access_id = 0 AND owner_guid = {$user->guid}";
+			$this->db->updateData($fasql);
+
+			$success_count++;
+		}
+	}
+
+	/**
+	 * Get id of current friends ACL id or create a new one if necessary
+	 *
+	 * @return int|bool $friends_acl_id The ACL id or false on error
+	 */
+	private function getFriendsAclId($user) {
+		$friends_acl = false;
+
+		if ($user->friends_acl) {
+			// Use existing ACL
+			$friends_acl = $this->accessCollections->get($user->friends_acl);
+		}
+
+		if (!$friends_acl) {
+			// Create a new ACL to hold the friends
+			$friends_acl_id = $this->accessCollections->create("acl:friends:{$user->guid}", $user->guid);
+		}
+
+		return $friends_acl_id;
+	}
+
+	/**
+	 * Get id of current friends ACL id or create a new one if necessary
+	 *
+	 * @return int|bool $friends_acl_id The ACL id or false on error
+	 */
+	private function getPrivateAclId(ElggUser $user) {
+		$private_acl = false;
+
+		if ($user->private_acl) {
+			$private_acl = $this->accessCollections->get($user->private_acl);
+		}
+
+		if (!$private_acl) {
+			$private_acl_id = $this->accessCollections->create("acl:private:{$user->guid}", $user->guid);
+		}
+
+		return $private_acl_id;
+	}
+
+	/**
+	 * Check if there are users without personal ACL or friends ACL
+	 *
+	 * @return boolean
+	 */
+	public function isRequired() {
+		$dbprefix = $this->db->getTablePrefix();
+
+		$friends_acl_id = $this->metastringsTable->getId('friends_acl');
+
+		// TODO Add a \Elgg\Database\AccessCollections::getFriendsAclId($user)
+		// method which returns false for non-existing ACL?
+		// @see comment: https://github.com/Elgg/Elgg/pull/8166#issuecomment-91807940
+		$friends_acl = $this->entityTable->getEntities(array(
+			'type' => 'user',
+			'wheres' => array(
+				"NOT EXISTS (
+			SELECT 1 FROM {$dbprefix}metadata md
+			WHERE md.entity_guid = e.guid
+				AND md.name_id = $friends_acl_id"
+			),
+			'count' => true
+		));
+
+		$private_acl_id = $this->metastringsTable->getId('private_acl');
+
+		// TODO Add a \Elgg\Database\AccessCollections::getPrivateAclId($user)
+		// method which returns false for non-existing ACL?
+		// @see comment: https://github.com/Elgg/Elgg/pull/8166#issuecomment-91807940
+		$private_acl = $this->entityTable->getEntities(array(
+			'type' => 'user',
+			'wheres' => array(
+				"NOT EXISTS (
+			SELECT 1 FROM {$dbprefix}metadata md
+			WHERE md.entity_guid = e.guid
+				AND md.name_id = $private_acl_id"
+			),
+			'count' => true
+		));
+
+		$return = $friends_acl && $private_acl;
```

Comment:
This should be an OR operator I think, if they have a friends acl but not a private acl we still need to run the upgrade